### PR TITLE
Refactor register app interface request logic

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Part of implementation #

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Tested with
with https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Smoke/API/037_RegisterAppInterface_PositiveCase_SUCCESS.lua 

### Summary
[Summary of PR changes]

##### Enhancements
Required for processing resumption error handing
Extracted some functionalities to functions

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 1]
```diff
@@ -89,6 +89,9 @@ class RegisterAppInterfaceRequest
    **/
   void SendRegisterAppInterfaceResponseToMobile();
 
+  void FillApplicationParams(
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan please provide some description for this function
BTW why it is in public section?

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 2]
```diff
@@ -105,7 +108,9 @@ class RegisterAppInterfaceRequest
    * considering application type
    * @param app_type Type of application
    **/
-  void SendRegisterAppInterfaceResponseToMobile(ApplicationType app_type);
+  void SendRegisterAppInterfaceResponseToMobile(ApplicationType app_type,
+      const std::__cxx11::string& add_info,
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan please rename this and provide description for a new params

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 3]
```diff
@@ -215,12 +219,28 @@ class RegisterAppInterfaceRequest
    */
   bool IsApplicationSwitched();
 
+  void TransformPolicyAppIdToUpper();
+
+  //  void ;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan please provide a description

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 4]
```diff
@@ -215,12 +219,28 @@ class RegisterAppInterfaceRequest
...
+  void StartResumptionDataRestore();
+
  private:
   std::string response_info_;
   mobile_apis::Result::eType result_code_;
 
   policy::PolicyHandlerInterface& GetPolicyHandler();
   DISALLOW_COPY_AND_ASSIGN(RegisterAppInterfaceRequest);
+  void WaitForHMIReady();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan please move all these functions up before members declaration

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 5]
```diff
@@ -218,87 +207,77 @@ void RegisterAppInterfaceRequest::Run() {
...
-  }
+  return true;
+}
 
-  if (IsWhiteSpaceExist()) {
-    LOG4CXX_INFO(logger_,
-                 "Incoming register app interface has contains \t\n \\t \\n");
-    SendResponse(false, mobile_apis::Result::INVALID_DATA);
-    return;
+void RegisterAppInterfaceRequest::IncrementDiplicateNameCounter(
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan `IncrementDuplicateNameCounter`

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 6]
```diff
@@ -405,8 +390,137 @@ void RegisterAppInterfaceRequest::Run() {
...
+  }
+  return DataResumeResult::RESUME_DATA;
+}
+
+void RegisterAppInterfaceRequest::Run() {
+  using namespace helpers;
+  LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_DEBUG(logger_, "Connection key is " << connection_key());
+
+  WaitForHMIReady();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan suggest to rename `WaitForHmiIsReady`

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 7]
```diff
@@ -405,8 +390,137 @@ void RegisterAppInterfaceRequest::Run() {
...
+
+  if (application_manager_.IsStopping()) {
+    LOG4CXX_WARN(logger_, "The ApplicationManager is stopping!");
+    return;
+  }
+
+  if (ProcessApplicationTransportSwitching()) {
+    return;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan suggest to add log messages for all `if` with `return` to understand what was the reason

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 8]
```diff
@@ -797,7 +872,7 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile(
 
 DEPRECATED void
 RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile() {
-  SendRegisterAppInterfaceResponseToMobile(ApplicationType::kNewApplication);
+  // SendRegisterAppInterfaceResponseToMobile(ApplicationType::kNewApplication);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan I think it should be uncommented

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 9]
```diff
@@ -1334,37 +1407,22 @@ bool RegisterAppInterfaceRequest::IsApplicationSwitched() {
...
+  const std::string policy_app_id = msg_params[strings::app_id].asString();
+  std::string new_policy_app_id = policy_app_id;
+  std::transform(policy_app_id.begin(),
+                 policy_app_id.end(),
+                 new_policy_app_id.begin(),
+                 ::tolower);
+  msg_params[strings::app_id] = new_policy_app_id;
+}
+
+void RegisterAppInterfaceRequest::StartResumptionDataRestore() {}
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan have you forgot to remove this one?

**Comment by LuxoftAKutsan** [comment]
@AKalinich-Luxoft all comments fixed

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 10]
```diff
@@ -90,6 +90,14 @@ class RegisterAppInterfaceRequest
   void SendRegisterAppInterfaceResponseToMobile();
 
  private:
+  /**
+   * @brief FillApplicationParams set app application attributes from the RAI
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan redundant word `app`

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 11]
```diff
@@ -105,7 +113,9 @@ class RegisterAppInterfaceRequest
    * considering application type
    * @param app_type Type of application
    **/
-  void SendRegisterAppInterfaceResponseToMobile(ApplicationType app_type);
+  void SendRegisterAppInterfaceResponseToMobile(ApplicationType app_type,
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan still can't see description for a new params

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 12]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
    */
   bool IsApplicationSwitched();
 
- private:
+  /**
+   * @brief TransformPolicyAppIdToUpper make policy app_id  in msg_params
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan redundant space in this line

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 13]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+   * uppercase.
+   * SDL compare policy appid in not case sensitive mode.
+   */
+  void TransformPolicyAppIdToUpper();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan as I could see from implementation you are transforming app id to lower. This function should be renamed

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 14]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+
+  /**
+   * @brief WaitForHMIIsReady blocking function. Waits for HMI be ready for
+   * requests processing
+   */
+  void WaitForHMIIsReady();
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan const?

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 15]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+
+  /**
+   * @brief IsApplicationForbidden check if application is in list of forbidden
+   * applications
+   * @return true if application os not allowed to register, othervise return
```
[the remark this answers is not in the document]
Comment (LuxoftAKutsan) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @AByzhynar fixed 900445f732

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 16]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+   * false.
+   */
+  bool IsApplicationForbidden();
```
[the remark this answers is not in the document]
Comment (LuxoftAKutsan) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @AByzhynar fixed 900445f732.
Thanks for the review in your vacation

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 17]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+
+  /**
+   * @brief SetupAppDeviceInfo add applicaiton device information to policies
+   * @param application applicaiton to process
+   */
+  void SetupAppDeviceInfo(
+      application_manager::ApplicationSharedPtr application);
+
+  /**
+   * @brief The DataResumeResult enum relust of resumption data check
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan typo

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h [hunk 18]
```diff
@@ -215,11 +224,76 @@ class RegisterAppInterfaceRequest
...
+   * RESUME_DATA : applicaiton data should be resumed
+   * NO_HASH : hash_id is ommited in RAI. No data resumption required
+   * WRONG_HASH : hash_id frim RAI does not match to savedhash id, resumption
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/commands/mobile/register_app_interface_request.h]: @LuxoftAKutsan typo

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 19]
```diff
@@ -218,87 +207,76 @@ void RegisterAppInterfaceRequest::Run() {
...
+  auto app_type = ApplicationType::kSwitchedApplicationWrongHashId;
+  if ((*message_)[strings::msg_params].keyExists(strings::hash_id)) {
+    const auto hash_id =
+        (*message_)[strings::msg_params][strings::hash_id].asString();
 
-  if (IsApplicationWithSameAppIdRegistered()) {
-    SendResponse(false, mobile_apis::Result::DISALLOWED);
-    return;
+    auto& resume_ctrl = application_manager_.resume_controller();
+    if (resume_ctrl.CheckApplicationHash(app, hash_id)) {
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan please add some logs to see if hash_id was found and the result of has_id check in resume_ctrl

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 20]
```diff
@@ -404,9 +388,147 @@ void RegisterAppInterfaceRequest::Run() {
...
+
+  if (IsApplicationForbidden()) {
+    LOG4CXX_WARN(logger_, "Application is Forbidden ");
+    SendResponse(false, mobile_apis::Result::TOO_MANY_PENDING_REQUESTS);
+    return;
+  }
+
+  if (IsApplicationWithSameAppIdRegistered()) {
+    LOG4CXX_WARN(logger_,
+                 "Application with the policy same app_id already registered ");
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan 
```
Application with the same policy app_id already registered 
```

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 21]
```diff
@@ -404,9 +388,147 @@ void RegisterAppInterfaceRequest::Run() {
...
+  if (mobile_apis::Result::SUCCESS != coincidence_result) {
+    LOG4CXX_ERROR(logger_, "Coincidence check failed.");
+    IncrementDuplicateNameCounter(coincidence_result);
+    SendResponse(false, coincidence_result);
+    return;
+  }
+
+  if (IsWhiteSpaceExist()) {
+    LOG4CXX_ERROR(logger_,
+                  "Incoming register app interface has contains tabulations or "
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan please remove has or contains word

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 22]
```diff
@@ -563,28 +685,45 @@ void FillUIRelatedFields(smart_objects::SmartObject& response_params,
...
+  const auto& msg_params = (*message_)[strings::msg_params];
+  if (msg_params[strings::language_desired].asInt() !=
+          hmi_capabilities_.active_vr_language() ||
+      msg_params[strings::hmi_display_language_desired].asInt() !=
+          hmi_capabilities_.active_ui_language()) {
+    LOG4CXX_WARN(logger_,
+                 "Wrong language on registering application "
+                     << application->name().c_str());
+
+    LOG4CXX_ERROR(
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan I think this message could be merged with the previous one

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 23]
```diff
@@ -760,18 +844,17 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile(
...
+      policy_handler_.AddApplication(application->policy_app_id(), hmi_types);
 
   response_params[strings::icon_resumed] =
       file_system::FileExists(application->app_icon_path());
 
-  SendResponse(true, result_code, add_info.c_str(), &response_params);
-  SendOnAppRegisteredNotificationToHMI(
-      *(application.get()), resumption, need_restore_vr);
+  SendResponse(true, result_code_, response_info_.c_str(), &response_params);
+  SendOnAppRegisteredNotificationToHMI(*(application.get()), need_restore_vr);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan can we use just `*application` ?

**Pull request merged**